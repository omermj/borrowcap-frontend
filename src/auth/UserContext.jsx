import React from "react";

/** Context to provide currentUser thoughout the app */

const UserContext = React.createContext();

export default UserContext;
