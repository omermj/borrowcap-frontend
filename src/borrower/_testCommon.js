import BorrowcapApi from "../api/api";

// Mock user
const testUser = {
  id: 1,
  username: "test",
  firstName: "test",
  lastName: "test",
  email: "testemail@email.com",
};

const commonBeforeEach = () => {
  // Mock api call with active requests
  BorrowcapApi.getActiveRequestsByBorrowerId.mockResolvedValue([
    {
      id: 4,
      amtRequested: 10000,
      purpose: "Business",
      appOpenDate: "2021/01/01",
      interestRate: 0.05,
      term: 12,
      installmentAmt: 1000,
    },
    {
      id: 5,
      amtRequested: 20000,
      purpose: "Home",
      appOpenDate: "2021/02/01",
      interestRate: 0.07,
      term: 24,
      installmentAmt: 2000,
    },
  ]);

  BorrowcapApi.getApprovedRequestsByUserId.mockResolvedValue({
    borrower: [
      {
        id: 2,
        amtRequested: "13000",
        amtApproved: "12000",
        amtFunded: "6000",
        purpose: "Education",
        appOpenDate: "2023-01-01",
        appApprovedDate: "2023/01/05",
        fundingDeadline: "2023/02/05",
        interestRate: "0.07",
        term: 24,
        installmentAmt: "537.27",
        availableForFunding: false,
      },
    ],
  });
};

module.exports = { testUser, commonBeforeEach };
