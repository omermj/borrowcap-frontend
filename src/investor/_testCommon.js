import BorrowcapApi from "../api/api";

// Mock user
const testUser = {
  id: 1,
  username: "test",
  firstName: "test",
  lastName: "test",
  email: "testemail@email.com",
};

const commonBeforeEach = async () => {
  // Mock api calls
  await BorrowcapApi.getFundedLoansByUserId.mockResolvedValue({
    investor: [
      {
        id: 1,
        investorId: 3,
        amtInvested: "5000",
        amtFunded: "5000",
        fundedDate: "2023/01/01",
        interestRate: "0.05",
        term: 12,
        installmentAmt: "428.04",
        remainingBalance: "5000",
      },
    ],
  });
  await BorrowcapApi.getAvailableInvestments.mockResolvedValue([
    {
      id: 7,
      amtApproved: "8000",
      amtFunded: "0",
      purpose: "Medical",
      approvedDate: "2023/01/01",
      fundingDeadline: "2023/02/1",
      interestRate: "0.07",
      term: 24,
    },
  ]);
  await BorrowcapApi.getApprovedRequestsByUserId.mockResolvedValue({
    investor: [
      {
        id: 6,
        amtApproved: "25000",
        amtPledged: "5000",
        approvedDate: "2023/01/01",
        fundingDeadline: "2023/02/01",
        interestRate: "0.07",
        term: 36,
      },
    ],
  });
  BorrowcapApi.getApprovedRequest.mockResolvedValue({
    id: 8,
    borrowerId: 2,
    amtRequested: "13000",
    amtApproved: "12000",
    amtFunded: "6000",
    purpose: "Education",
    purposeId: 3,
    appOpenDate: "2023-01-01",
    appApprovedDate: "2023/01/05",
    fundingDeadline: "2023/02/05",
    interestRate: "0.07",
    term: 24,
    installmentAmt: "537.27",
    availableForFunding: true,
    isFunded: false,
    annualIncome: "100000",
    otherMonthlyDebt: "1200",
  });
};

module.exports = { testUser, commonBeforeEach };
