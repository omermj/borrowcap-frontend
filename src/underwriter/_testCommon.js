import BorrowcapApi from "../api/api";

// Mock user
const testUser = {
  id: 1,
  username: "test",
  firstName: "test",
  lastName: "test",
  email: "testemail@email.com",
};

const commonBeforeEach = async () => {
  // Mock api calls
  await BorrowcapApi.getLoanRequests.mockResolvedValue([
    {
      id: 2,
      borrowerId: 2,
      amtRequested: "10000",
      purpose: "Education",
      purposeId: 3,
      appOpenDate: "2023/01/01",
      interestRate: "0.07",
      term: 36,
      installmentAmt: "319.86",
    },
  ]);

  await BorrowcapApi.getApprovedRequests.mockResolvedValue([
    {
      id: 1,
      borrowerId: 2,
      amtRequested: "10000",
      amtApproved: "9000",
      amtFunded: "5000",
      purpose: "Car",
      purposeId: 2,
      appOpenDate: "2023/01/01",
      appApprovedDate: "2023/01/05",
      fundingDeadline: "2023/02/05",
      interestRate: "0.05",
      term: 12,
      installmentAmt: "428.04",
      availableForFunding: false,
      isFunded: true,
    },
  ]);
};

module.exports = { testUser, commonBeforeEach };
