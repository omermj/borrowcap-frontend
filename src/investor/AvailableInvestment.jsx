import { Container, Row, Col } from "react-bootstrap";
import { useState, useEffect, useContext } from "react";
import { useParams, useNavigate } from "react-router-dom";
import BorrowcapApi from "../api/api";
import LoadingSpinner from "../common/LoadingSpinner";
import { formatCurrency, formatDate, formatPercent } from "../helpers/format";
import { Formik } from "formik";
import { Form, Button } from "react-bootstrap";
import UserContext from "../auth/UserContext";

/** Displays a single Available Investment */

const AvailableInvestment = () => {
  const { currentUser } = useContext(UserContext);
  const { id } = useParams();
  const [data, setData] = useState(null);
  const navigate = useNavigate();
  // const [infoLoaded, setInfoLoaded] = useState(false);

  useEffect(() => {
    const getInvestment = async () => {
      const result = await BorrowcapApi.getApprovedRequest(id);
      if (result) {
        setData(result);
        // setInfoLoaded(true);
      }
    };
    // setInfoLoaded(false);
    getInvestment();
  }, [id]);

  if (!data) return <LoadingSpinner />;

  return (
    <div>
      <h3 className="text-center mb-4">Investment Details</h3>
      <Container>
        <Row className="mb-4">
          <Col>
            <div>Amount Requested: {formatCurrency(data.amtRequested)}</div>
            <div>Amount Approved: {formatCurrency(data.amtApproved)}</div>
            <div>Amount Funded: {formatCurrency(data.amtFunded)}</div>
            <div>Loan Purpose: {data.purpose} </div>
            <div>Application Open Date: {formatDate(data.appOpenDate)}</div>
            <div>
              Application Approval Date: {formatDate(data.appApprovedDate)}
            </div>
          </Col>
          <Col>
            <div>Funding Deadline: {formatDate(data.fundingDeadline)} </div>
            <div>Term: {data.term} months </div>
            <div>Installment Amount: {formatCurrency(data.installmentAmt)}</div>
            <div>
              Borrower&apos;s Income: {formatCurrency(data.annualIncome)}
            </div>
            <div>
              Other Monthly Debt: {formatCurrency(data.otherMonthlyDebt)}
            </div>
          </Col>
        </Row>
        <Row>
          <Col className="form-wrapper">
            <div className="form-inner">
              <Formik
                initialValues={{
                  amount: "",
                }}
                onSubmit={async (values, { setSubmitting, setErrors }) => {
                  values.investorId = currentUser.id;
                  const res = await BorrowcapApi.fundApprovedRequest(
                    id,
                    values
                  );
                  if (!!res) navigate("/investor");
                }}
              >
                {({
                  values,
                  errors,
                  handleChange,
                  handleSubmit,
                  isSubmitting,
                }) => (
                  <Form onSubmit={handleSubmit}>
                    <Form.Group className="mb-3" controlId="investFormAmount">
                      <Form.Label>Amount</Form.Label>
                      <Form.Control
                        type="number"
                        name="amount"
                        placeholder={`Max amount: ${formatCurrency(
                          data.amtApproved - data.amtFunded
                        )}`}
                        min={0}
                        value={values.amount}
                        onChange={handleChange}
                        required
                      />
                    </Form.Group>
                    <div className="text-center mt-4">
                      <Button
                        variant="primary"
                        type="submit"
                        disabled={isSubmitting}
                        className="px-4 btn-dark"
                      >
                        Invest
                      </Button>
                    </div>
                  </Form>
                )}
              </Formik>
            </div>
          </Col>
        </Row>
      </Container>
    </div>
  );
};

export default AvailableInvestment;
