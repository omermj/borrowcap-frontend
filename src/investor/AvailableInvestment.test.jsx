import { render, waitFor, screen, act } from "@testing-library/react";
import AvailableInvestment from "./AvailableInvestment.jsx";
import { expect, it } from "vitest";
import UserContext from "../auth/UserContext.jsx";
import { commonBeforeEach, testUser } from "./_testCommon.js";
import BorrowcapApi from "../api/api.js";
import { BrowserRouter } from "react-router-dom";

// Mock api calls
vi.mock("../api/api");
beforeEach(commonBeforeEach);

// Smoke test
it("renders without crashing", () => {
  render(
    <BrowserRouter>
      <UserContext.Provider value={{ currentUser: testUser }}>
        <AvailableInvestment />
      </UserContext.Provider>
    </BrowserRouter>
  );
});

// Snapshot test
it("matches snapshot", () => {
  const { asFragment } = render(
    <BrowserRouter>
      <UserContext.Provider value={{ currentUser: testUser }}>
        <AvailableInvestment />
      </UserContext.Provider>
    </BrowserRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});

// Test Available Investment data
it("renders information properly", async () => {
  render(
    <BrowserRouter>
      <UserContext.Provider value={{ currentUser: testUser }}>
        <AvailableInvestment />
      </UserContext.Provider>
    </BrowserRouter>
  );

  // Wait for api to resolve
  await waitFor(async () => {
    expect(BorrowcapApi.getApprovedRequest).toHaveBeenCalled();
  });

  // Verify approved request information is rendered properly
  expect(screen.getByText("Amount Requested: $13,000.00")).toBeInTheDocument();
  expect(screen.getByText("Amount Approved: $12,000.00")).toBeInTheDocument();
  expect(screen.getByText("Amount Funded: $6,000.00")).toBeInTheDocument();
  expect(screen.getByText("Loan Purpose: Education")).toBeInTheDocument();

  // Verify buttons rendered properly
  expect(screen.getByText("Invest")).toBeInTheDocument();
});

// Test Invest button click
it("handles Enable Funding button click", async () => {
  render(
    <BrowserRouter>
      <UserContext.Provider value={{ currentUser: testUser }}>
        <AvailableInvestment />
      </UserContext.Provider>
    </BrowserRouter>
  );

  // Wait for api to resolve
  await waitFor(() => {
    expect(BorrowcapApi.getApprovedRequest).toHaveBeenCalled();
  });

  // Click enable funding button
  async () => {
    const investButton = screen.getByText("Invest");
    fireEvent.click(investButton);

    await waitFor(() => {
      expect(BorrowcapApi.fundApprovedRequest).toHaveBeenCalled();
    });
  };
});
