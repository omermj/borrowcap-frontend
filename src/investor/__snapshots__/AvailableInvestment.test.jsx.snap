// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="LoadingSpinner"
  >
     Loading...
  </div>
</DocumentFragment>
`;
